File: ios/Shared/Backend/MailModels.swift
import Foundation

struct MailMessageSummary: Decodable, Identifiable, Hashable {
    var uid: String
    var subject: String
    var fromName: String?
    var fromAddress: String?
    var receivedAt: String?
    var preview: String?
    var isUnread: Bool

    var id: String { uid }

    enum CodingKeys: String, CodingKey {
        case uid, subject, preview
        case fromName = "from_name"
        case fromAddress = "from_address"
        case receivedAt = "received_at"
        case isUnread = "is_unread"
    }
}

struct MailInboxSummary: Decodable {
    var account: String
    var mailbox: String
    var unreadCount: Int
    var messages: [MailMessageSummary]

    enum CodingKeys: String, CodingKey {
        case account, mailbox, messages
        case unreadCount = "unread_count"
    }
}

struct MailboxSummary: Decodable, Identifiable, Hashable {
    var name: String
    var label: String
    var specialUse: String?
    var messageCount: Int?
    var unreadCount: Int?

    var id: String { name }

    enum CodingKeys: String, CodingKey {
        case name, label
        case specialUse = "special_use"
        case messageCount = "message_count"
        case unreadCount = "unread_count"
    }
}

struct MailMessageDetail: Decodable, Identifiable {
    var uid: String
    var mailbox: String
    var subject: String
    var fromName: String?
    var fromAddress: String?
    var toRecipients: [String]
    var ccRecipients: [String]
    var receivedAt: String?
    var preview: String?
    var bodyText: String?
    var attachmentNames: [String]
    var isUnread: Bool

    var id: String { "\(mailbox):\(uid)" }

    enum CodingKeys: String, CodingKey {
        case uid, mailbox, subject, preview
        case fromName = "from_name"
        case fromAddress = "from_address"
        case toRecipients = "to_recipients"
        case ccRecipients = "cc_recipients"
        case receivedAt = "received_at"
        case bodyText = "body_text"
        case attachmentNames = "attachment_names"
        case isUnread = "is_unread"
    }
}

struct MailMessageSelection: Hashable {
    var uid: String
    var mailbox: String
    var subject: String
}

File: package/src/tue_api_wrapper/mail_models.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class MailMessageSummary:
    uid: str
    subject: str
    from_name: str | None
    from_address: str | None
    received_at: str | None
    preview: str | None
    is_unread: bool
    is_approved_broadcast: bool = False


@dataclass(frozen=True)
class MailInboxSummary:
    account: str
    mailbox: str
    unread_count: int
    messages: tuple[MailMessageSummary, ...]


@dataclass(frozen=True)
class MailboxSummary:
    name: str
    label: str
    special_use: str | None
    message_count: int | None
    unread_count: int | None


@dataclass(frozen=True)
class MailMessageDetail:
    uid: str
    mailbox: str
    subject: str
    from_name: str | None
    from_address: str | None
    to_recipients: tuple[str, ...]
    cc_recipients: tuple[str, ...]
    received_at: str | None
    preview: str | None
    body_text: str | None
    attachment_names: tuple[str, ...]
    is_unread: bool
    is_approved_broadcast: bool = False
